software/contrib/quantizer.py

```diff
@@ -0,0 +1,537 @@
+            
+            if self.active_screen == self.kb:
+                self.active_screen = self.menu
+            else:
+                self.active_screen = self.kb
+        
+    def load(self):
+        """Load the persistent settings from storage and apply them
+        """
+        state = self.load_state_json()
+        
+        loaded_scale = state.get("scale", [True]*12)  # default to a chromatic scale
+        self.scale.notes = loaded_scale
+        
+        self.root = state.get("root", self.root)
+        self.octave = state.get("octave", self.octave)
+        self.intervals = state.get("intervals", self.intervals)
+        self.mode = state.get("mode", self.mode)
+    
+    def save(self):
+        """Save the current settings to persistent storage
+        """
+        state = {
+            "scale": self.scale.notes,
+            "root": self.root,
+            "octave": self.octave,
+            "intervals": self.intervals,
+            "mode": self.mode
+        }
+        self.save_state_json(state)
+        
+    @classmethod
+    def display_name(cls):
+        return "Quantizer"
+       
+    def quantize(self, analog_in):
+        """Take an analog signal and process it
+
+        Sets self.current_note and self.output_voltage
+        """
+        
+        (volts, semitone) = self.scale.quantize(analog_in)
+        
+        # apply our octave & transposition offsets
+        volts = volts + self.octave * VOLTS_PER_OCTAVE + self.root * VOLTS_PER_SEMITONE
+        
+        self.output_voltage = volts
+        self.current_note = semitone
+       
+    def read_quantize_output(self):
+        """Read the input signal, quantize it, set outputs 1-5 accordingly
+
+        Called by the main loop in continuous mode or the rising clock handler
+        in triggered mode
+        """
+        self.input_voltage = ain.read_voltage(128)
+        self.quantize(self.input_voltage)
+        
+        cv1.voltage(self.output_voltage)
+        
+        for i in range(len(self.aux_outs)):
+            self.aux_outs[i].voltage(self.output_voltage + self.intervals[i] * VOLTS_PER_SEMITONE)
+    
+    def main(self):
+        """The main loop; reads from ain, sets the output voltages
+        """
+        
+        while True:
+            # Update at 100Hz
+            CYCLE_RATE = 0.01
+            time.sleep(CYCLE_RATE)
+            
+            # Check if we've been idle for too long; if so, blank the screen
+            # to prevent burn-in
+            now = time.ticks_ms()
+            if time.ticks_diff(now, self.last_interaction_time) > SCREENSAVER_TIMEOUT_MS:
+                self.active_screen = self.screensaver
+            
+            # read the encoder value from knob 1 and apply it to the
+            # active menu items for the UI
+            self.highlight_note = round(knob_rescale(k1, 0, len(self.scale)-1))
+            self.menu_item = round(knob_rescale(k1, 0, len(self.menu.menu_items)-1))
```

Comment:
[question] these two lines cause k1 to be read twice, when only one value will be used on any particular trip through the loop (depending on which screen is being viewed). Why do the reads here instead of where they are being used, as you do with knob 2? I feel like i am missing something.

Reply:
That was a holdover from an earlier version where more work was being done inside the main loop. Things got re-organized, and that part just got ignored because it was working.

I've relocated reading k1 now